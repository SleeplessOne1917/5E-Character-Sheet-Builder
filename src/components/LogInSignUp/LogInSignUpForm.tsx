import Button, { ButtonType } from '../Button/Button';
import { Formik, FormikHelpers } from 'formik';

import Link from 'next/link';
import TextInput from '../TextInput/TextInput';
import classes from './LogInSignUpForm.module.css';

type LogInSignUpValues = {
	username: string;
	password: string;
	email?: string;
};

export type LogInSignUpProps = {
	onSubmit: (
		values: LogInSignUpValues,
		formikHelpers: FormikHelpers<LogInSignUpValues>
	) => void | Promise<any>;
	schema: any;
	type: 'logIn' | 'signUp';
};

const LogInSignUpForm = ({ type, schema, onSubmit }: LogInSignUpProps) => (
	<Formik
		initialValues={{ username: '', password: '' }}
		validationSchema={schema}
		onSubmit={onSubmit}
	>
		{({
			values,
			handleChange,
			handleBlur,
			handleSubmit,
			isSubmitting,
			errors,
			touched,
			setTouched
		}) => (
			<form onSubmit={handleSubmit} className={classes.form} data-testid="form">
				<TextInput
					touched={touched.username}
					error={errors.username}
					id="username"
					label="Username"
					onBlur={handleBlur}
					onChange={event => {
						setTouched({ ...touched, username: false });
						handleChange(event);
					}}
					value={values.username}
				/>
				<TextInput
					touched={touched.password}
					error={errors.password}
					id="password"
					label="Password"
					onBlur={handleBlur}
					onChange={event => {
						setTouched({ ...touched, password: false });
						handleChange(event);
					}}
					value={values.password}
					type={type === 'logIn' ? 'password' : 'validate-password'}
				/>
				{type === 'signUp' && (
					<>
						<TextInput
							touched={touched.email}
							error={errors.email}
							id="email"
							label={
								<>
									Email{' '}
									<span className={classes['optional-text']}>(optional)</span>
								</>
							}
							onBlur={handleBlur}
							onChange={event => {
								setTouched({ ...touched, email: false });
								handleChange(event);
							}}
							value={values?.email ?? ''}
						/>
						<p className={classes['email-blurb']}>
							Forgetful? Providing your email will allow you to request a
							password reset or a reminder of your username. It will be
							encrypted so your data will still be secure in the event of a data
							breach. If you don&apos;t provide an email, make sure to put extra
							points into INT to help you with your remember credentials checks!
						</p>
					</>
				)}
				<Button
					disabled={isSubmitting}
					type={ButtonType.submit}
					positive
					size="large"
					spacing={4}
				>
					{type === 'logIn' ? 'Log In' : 'Sign Up'}
				</Button>
				{type === 'logIn' && (
					<Link href="/forgot">
						<a className={classes['forgot-link']}>
							Forgot your username or password?
						</a>
					</Link>
				)}
			</form>
		)}
	</Formik>
);

export default LogInSignUpForm;
