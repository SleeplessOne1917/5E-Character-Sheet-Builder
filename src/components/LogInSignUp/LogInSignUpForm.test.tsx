import '@testing-library/jest-dom';

import * as stories from './LogInSignUpForm.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/react';
import userEvent from '@testing-library/user-event';

const { LogIn, SignUp } = composeStories(stories);

describe('renders correctly', () => {
	it('when log in', () => {
		render(<LogIn />);

		expect(screen.getByTestId('form')).toMatchSnapshot();
	});

	it('when sign up', () => {
		render(<SignUp />);

		expect(screen.getByTestId('form')).toMatchSnapshot();
	});
});

it('toggles show password when eye icon is pressed', async () => {
	render(<SignUp />);

	const passwordField = screen.getByLabelText(
		/^Password$/i
	) as HTMLInputElement;
	expect(screen.queryByLabelText(/Hide Password/i)).not.toBeInTheDocument();
	expect(passwordField.type).toBe('password');

	await userEvent.click(screen.getByLabelText(/Show Password/i));
	expect(screen.queryByLabelText(/Show Password/i)).not.toBeInTheDocument();
	expect(passwordField.type).toBe('text');

	await userEvent.click(screen.getByLabelText(/Hide Password/i));
	expect(screen.queryByLabelText(/Hide Password/i)).not.toBeInTheDocument();
	expect(passwordField.type).toBe('password');
});

describe('error text', () => {
	it('only shows for username after blur', async () => {
		render(<LogIn />);

		expect(screen.queryByText(/Username is required/i)).not.toBeInTheDocument();

		await userEvent.tab();
		await userEvent.tab();
		await userEvent.tab();

		expect(screen.getByText(/Username is required/i)).toBeInTheDocument();
	});

	it('only shows for password after blur', async () => {
		render(<LogIn />);

		expect(screen.queryByText(/Password is required/i)).not.toBeInTheDocument();

		await userEvent.tab();
		await userEvent.tab();
		await userEvent.tab();

		expect(screen.getByText(/Password is required/i)).toBeInTheDocument();
	});

	it('only shows for email after blur', async () => {
		render(<SignUp />);

		expect(
			screen.queryByText(/Enter your email in the format/i)
		).not.toBeInTheDocument();

		await userEvent.type(screen.getByLabelText(/Email/i), 'notAnEmail');
		await userEvent.tab();

		expect(
			screen.getByText(/Enter your email in the format/i)
		).toBeInTheDocument();
	});
});
