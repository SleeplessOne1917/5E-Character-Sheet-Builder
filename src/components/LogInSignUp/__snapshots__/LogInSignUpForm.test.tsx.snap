// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when log in 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div
    class="input-and-error-container"
    data-testid="text-input"
  >
    <div
      class="input-and-label-container"
    >
      <div
        class="input-container"
      >
        <input
          class="input"
          id="username"
          name="username"
          type="text"
          value=""
        />
      </div>
      <label
        class="label"
        for="username"
      >
        Username
      </label>
    </div>
  </div>
  <div
    class="input-and-error-container"
    data-testid="text-input"
  >
    <div
      class="input-and-label-container"
    >
      <div
        class="input-container"
      >
        <input
          class="input"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <svg
          aria-hidden="false"
          aria-label="Show Password"
          class="eye"
          fill="currentColor"
          tabindex="0"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 12.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5z"
          />
          <path
            clip-rule="evenodd"
            d="M.664 10.59a1.651 1.651 0 010-1.186A10.004 10.004 0 0110 3c4.257 0 7.893 2.66 9.336 6.41.147.381.146.804 0 1.186A10.004 10.004 0 0110 17c-4.257 0-7.893-2.66-9.336-6.41zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <label
        class="label"
        for="password"
      >
        Password
      </label>
    </div>
  </div>
  <button
    class="button positive"
    style="font-size: 2.5rem; padding: 1rem; margin: 1rem;"
    type="submit"
  >
    Log In
  </button>
  <a
    class="forgot-link"
    href="/forgot"
  >
    Forgot your username or password?
  </a>
</form>
`;

exports[`renders correctly when sign up 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div
    class="input-and-error-container"
    data-testid="text-input"
  >
    <div
      class="input-and-label-container"
    >
      <div
        class="input-container"
      >
        <input
          class="input"
          id="username"
          name="username"
          type="text"
          value=""
        />
      </div>
      <label
        class="label"
        for="username"
      >
        Username
      </label>
    </div>
  </div>
  <div
    class="input-and-error-container"
    data-testid="text-input"
  >
    <div
      class="input-and-label-container"
    >
      <div
        class="input-container"
      >
        <input
          class="input"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <svg
          aria-hidden="false"
          aria-label="Show Password"
          class="eye"
          fill="currentColor"
          tabindex="0"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 12.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5z"
          />
          <path
            clip-rule="evenodd"
            d="M.664 10.59a1.651 1.651 0 010-1.186A10.004 10.004 0 0110 3c4.257 0 7.893 2.66 9.336 6.41.147.381.146.804 0 1.186A10.004 10.004 0 0110 17c-4.257 0-7.893-2.66-9.336-6.41zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <label
        class="label"
        for="password"
      >
        Password
      </label>
    </div>
    <div
      class="container"
      data-testid="container"
    >
      <div
        class="validator no-match"
      >
        <svg
          aria-hidden="true"
          class="icon x"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
            fill-rule="evenodd"
          />
        </svg>
         
        At least 8 characters long
      </div>
      <div
        class="validator no-match"
      >
        <svg
          aria-hidden="true"
          class="icon x"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
            fill-rule="evenodd"
          />
        </svg>
         
        At least 1 lowercase letter
      </div>
      <div
        class="validator no-match"
      >
        <svg
          aria-hidden="true"
          class="icon x"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
            fill-rule="evenodd"
          />
        </svg>
         
        At least 1 uppercase letter
      </div>
      <div
        class="validator no-match"
      >
        <svg
          aria-hidden="true"
          class="icon x"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
            fill-rule="evenodd"
          />
        </svg>
         
        At least 1 number
      </div>
      <div
        class="validator no-match"
      >
        <svg
          aria-hidden="true"
          class="icon x"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.97 3.97a.75.75 0 011.06 0L12 10.94l6.97-6.97a.75.75 0 111.06 1.06L13.06 12l6.97 6.97a.75.75 0 11-1.06 1.06L12 13.06l-6.97 6.97a.75.75 0 01-1.06-1.06L10.94 12 3.97 5.03a.75.75 0 010-1.06z"
            fill-rule="evenodd"
          />
        </svg>
         
        At least 1 special character
      </div>
    </div>
  </div>
  <div
    class="input-and-error-container"
    data-testid="text-input"
  >
    <div
      class="input-and-label-container"
    >
      <div
        class="input-container"
      >
        <input
          class="input"
          id="email"
          name="email"
          type="text"
          value=""
        />
      </div>
      <label
        class="label"
        for="email"
      >
        Email
         
        <span
          class="optional-text"
        >
          (optional)
        </span>
      </label>
    </div>
  </div>
  <p
    class="email-blurb"
  >
    Forgetful? Providing your email will allow you to request a password reset or a reminder of your username. It will be encrypted so your data will still be secure in the event of a data breach. If you don't provide an email, make sure to put extra points into INT to help you with your remember credentials checks!
  </p>
  <button
    class="button positive"
    style="font-size: 2.5rem; padding: 1rem; margin: 1rem;"
    type="submit"
  >
    Sign Up
  </button>
</form>
`;
